src/v_fmt.c

```diff
@@ -315,63 +315,289 @@ static void *DummyFlat(int lump, pu_tag tag)
     return lumpcache[lump];
 }
 
+typedef struct
+{
+    spng_ctx *ctx;
+    byte *image;
+    size_t image_size;
+    int width;
+    int height;
+    int color_key;
+} png_t;
+
 // Set memory usage limits for storing standard and unknown chunks,
 // this is important when reading untrusted files!
 #define PNG_MEM_LIMIT (1024 * 1024 * 64)
 
-patch_t *V_CachePatchNum(int lump, pu_tag tag)
+static boolean InitPNG(png_t *png, void *buffer, int buffer_length)
 {
-    if (lump >= numlumps)
+    spng_ctx *ctx = spng_ctx_new(0);
+
+    // Ignore and don't calculate chunk CRC's
+    int ret = spng_set_crc_action(ctx, SPNG_CRC_USE, SPNG_CRC_USE);
+
+    if (ret)
     {
-        I_Error("V_CachePatchNum: %d >= numlumps", lump);
+        I_Printf(VB_ERROR, "InitPNG: spng_set_crc_action %s\n",
+                 spng_strerror(ret));
+        return false;
     }
 
-    if (lumpcache[lump])
+    ret = spng_set_chunk_limits(ctx, PNG_MEM_LIMIT, PNG_MEM_LIMIT);
+
+    if (ret)
     {
-        Z_ChangeTag(lumpcache[lump], tag);
-        return lumpcache[lump];
+        I_Printf(VB_ERROR, "InitPNG: spng_set_chunk_limits %s\n",
+                 spng_strerror(ret));
+        return false;
     }
 
-    void *buffer = W_CacheLumpNum(lump, tag);
-    int buffer_length = W_LumpLength(lump);
+    ret = spng_set_png_buffer(ctx, buffer, buffer_length);
 
-    if (buffer_length < 8 || memcmp(buffer, "\211PNG\r\n\032\n", 8))
+    if (ret)
     {
-        return buffer;
+        I_Printf(VB_ERROR, "InitPNG: spng_set_png_buffer %s\n",
+                 spng_strerror(ret));
+        return false;
     }
 
-    byte *image = NULL, *translate = NULL;
+    png->ctx = ctx;
 
-    spng_ctx *ctx = spng_ctx_new(0);
+    return true;
+}
 
-    // Ignore and don't calculate chunk CRC's
-    spng_set_crc_action(ctx, SPNG_CRC_USE, SPNG_CRC_USE);
+static void FreePNG(png_t *png)
+{
+    spng_ctx_free(png->ctx);
+    if (png->image)
+    {
+        free(png->image);
+    }
+}
 
-    spng_set_chunk_limits(ctx, PNG_MEM_LIMIT, PNG_MEM_LIMIT);
+static boolean DecodePNG(png_t *png)
+{
+    struct spng_ihdr ihdr = {0};
+    int ret = spng_get_ihdr(png->ctx, &ihdr);
 
-    spng_set_option(ctx, SPNG_KEEP_UNKNOWN_CHUNKS, 1);
+    if (ret)
+    {
+        I_Printf(VB_ERROR, "DecodeImage: spng_get_ihdr %s\n",
+                 spng_strerror(ret));
+        return false;
+    }
 
-    spng_set_png_buffer(ctx, buffer, buffer_length);
+    png->width = ihdr.width;
+    png->height = ihdr.height;
 
-    struct spng_ihdr ihdr = {0};
-    int ret = spng_get_ihdr(ctx, &ihdr);
+    int fmt;
+    switch (ihdr.color_type)
+    {
+        case SPNG_COLOR_TYPE_INDEXED:
+            fmt = SPNG_FMT_PNG;
+            break;
+        case SPNG_COLOR_TYPE_GRAYSCALE:
```

Comment:
Is this fall through intended?

Reply:
Yes, I convert everything to `SPNG_FMT_RGB8` or `SPNG_FMT_RGBA8` for simplicity.